spec.md

```diff
@@ -1090,6 +1084,252 @@ message ControllerServiceCapability {
+  // Use cases for opaque parameters:
+  // - Specify a policy to automatically clean up the snapshot.
+  // - Specify an expiration date for the snapshot.
+  // - Specify whether the snapshot is readonly or read/write.
+  // - Specify if the snapshot should be replicated to some place.
+  // - Specify primary or secondary for replication systems that
+  //   support snapshotting only on primary.
+  map<string, string> parameters = 4;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
+  // SnapshotSize is the disk size of the snapshot. The purpose of this
+  // field is to give CO guidance on how much space is needed to create
+  // a volume from this snapshot. The size of the volume must not be
+  // less than the size of the source snapshot. This field is OPTIONAL.
+  // If SnapshotSize is not set, it indicates that this size is unknown.
+  SnapshotSize size = 1;
+
+  // Contains identity information for the created snapshot. This field
+  // is REQUIRED. The identity information will be used by the CO in
+  // subsequent calls to refer to the provisioned snapshot.
+  string id = 2;
+
+  // Identity information for the source volume. Note that create
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string source_volume_id = 3;
+
+  // Timestamp when the point-in-time snapshot is taken on the storage
+  // system. The format of this field should be a UnixNano time encoded
+  // as an int64. This field is REQUIRED.
+  int64 created_at = 4;
+
+  // The status of a snapshot.
+  SnapshotStatus status = 5;
+}
+
+message SnapshotSize {
+  // The disk size of the snapshot. The purpose of this field is to
+  // give CO guidance on how much space is needed to create a volume
+  // from this snapshot. The size of the volume must not be less than
+  // the size of the source snapshot. This field is OPTIONAL. The value
+  // of this field MUST NOT be negative.
+  int64 size_bytes = 1;
+}
+
+// The status of a snapshot. A snapshot could have the following
+// statuses:
+// UPLOADING - A snapshot is cut and is now being uploaded.
+// READY - A snapshot is ready for use.
+// Some cloud providers will upload the snapshot to a location in
+// the cloud (i.e., an object store) after the snapshot is cut.
+// Uploading is a long process that could take hours. At this phase,
+// since the snapshot is already cut, a `thaw` operation can be
+// performed so application can start running again (assuming a
+// `freeze` operation was done before taking the snapshot). The
+// status of the snapshot will become `READY` after the upload is
+// complete.
+// For cloud providers and storage systems that don't have the
```

Comment:
I LOVE the detail in these paragraphs. I think it might be more appropriate to move the bulk of this text into the markdown section of the spec and keep the commentary for the protos a bit more concise. Perhaps a better home would be the "RPC Interactions" section?

Reply:
Sure.